receiver/kafkametricsreceiver/consumer_scraper.go

```diff
@@ -45,31 +44,37 @@ func (s *consumerScraper) Name() string {
 	return consumersScraperName
 }
 
-func (s *consumerScraper) start(context.Context, component.Host) error {
+func (s *consumerScraper) setupClient() error {
+	if s.client != nil {
+		return nil
+	}
 	client, err := newSaramaClient(s.config.Brokers, s.saramaConfig)
 	if err != nil {
-		return fmt.Errorf("failed to create client while starting consumer scraper: %w", err)
+		return fmt.Errorf("failed to create client in consumer scraper: %w", err)
 	}
 	clusterAdmin, err := newClusterAdmin(s.config.Brokers, s.saramaConfig)
 	if err != nil {
 		if client != nil {
 			_ = client.Close()
 		}
-		return fmt.Errorf("failed to create cluster admin while starting consumer scraper: %w", err)
+		return fmt.Errorf("failed to create cluster admin in consumer scraper: %w", err)
 	}
 	s.client = client
 	s.clusterAdmin = clusterAdmin
 	return nil
 }
 
 func (s *consumerScraper) shutdown(_ context.Context) error {
-	if !s.client.Closed() {
+	if s.client != nil && !s.client.Closed() {
 		return s.client.Close()
 	}
 	return nil
 }
 
 func (s *consumerScraper) scrape(context.Context) (pmetric.Metrics, error) {
+	if err := s.setupClient(); err != nil {
+		return pmetric.Metrics{}, err
+	}
```

Comment:
This way of code reusing seems a bit confusing to me. It sounds like we setup a client on each scrape. If we remove the function it becomes much cleaner IMO

```suggestion
	if s.client != nil {
		s.client, err := newSaramaClient(s.config.Brokers, s.saramaConfig)
		if err != nil {
			return pmetric.Metrics{}, fmt.Errorf("failed to create client in consumer scraper: %w", err)
		}
	}
```

Another approach to avoid copying this code in all the scrapers would be using a wrapper for the scrape function, then each scraper creation would be:

```golang
scraperhelper.NewScraper(
		s.Name(),
		withClientSetup(s.scrape),
		scraperhelper.WithShutdown(s.shutdown),
	)
```

Reply:
I liked both of your suggestions, but went with the first (inlining) as it makes the code slightly easier to follow.